contracts/utils/NoncesSemiAbstracted.sol

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {Nonces} from "./Nonces.sol";
+
+/**
+ * @dev Alternative to {Nonces}, that support key-ed nonces.
+ *
+ * Follows the https://eips.ethereum.org/EIPS/eip-4337#semi-abstracted-nonce-support[ERC-4337's semi-abstracted nonce system].
+ */
+abstract contract NoncesSemiAbstracted is Nonces {
```

Comment:
It seems risky to completely override the logic of `Nonces`.

One possible issue would be an upgrade that adds this contract to one that already had `Nonces`. In that case the value of `nonces(owner)` would reset.

A conservative implementation could reuse the built-in mapping in `Nonces` when the key is 0.

Reply:
The overrides for `nonces(address)`, `_useNonce(address)` could be removed since the logic would remain the same.